### PR TITLE
2097679: Fixed script hang in non-interactive execution

### PR DESCRIPTION
**Pull request by DuckBoss**
- Resolved the script hanging issue caused by the script waiting on user inputs for information not specified through command line arguments by checking if the session is interactive and raising an error detailing the missing command line argument if not.
- Updated the man page "commands and options" section with a note to document this behavior.
- BZ: https://bugzilla.redhat.com/show_bug.cgi?id=2097679
- Card ID: ENT-5117

### CODE REVIEW

File: man/subscription-manager.8 [hunk 1]
```diff
@@ -58,6 +58,10 @@ command.
 .B subscription-manager
 has specific options available for each command, depending on what operation is being performed. Subscription Manager commands are related to the different subscription operations:
 
+.PP
+.B Note:
+Please note that using commands that requires providing a username, password, organization, or environments will require them to be passed as system arguments in a non-interactive session.
```
Comment (ptoscano) [inline man/subscription-manager.8:63]: s/requires/require/ (it refers to "commands", which is plural)

Comment (ptoscano) [inline man/subscription-manager.8:63]: i'd mention explicitly what are the affected options, e.g. "... providing an username using .B --username etc" (take care of the groff formatting); this makes it easier to see them, and to look/search for them

File: src/subscription_manager/cli_command/register.py [hunk 2]
```diff
@@ -387,6 +388,9 @@ def _prompt_for_environment(self):
         """
         By breaking this code out, we can write cleaner tests
         """
+        if not self._is_atty():
+            raise system_exit(os.EX_USAGE, "Error: --environments parameter not provided!")
```
Comment (ptoscano) [inline src/subscription_manager/cli_command/register.py:392]: `system_exit()` does not raise an exception, but directly causes sub-man to quit; hence, "raise" is not needed

Comment (ptoscano) [inline src/subscription_manager/cli_command/register.py:392]: error messages shown to the user must be translatable; use the `gettext()` function, also usually available as `_()`

Comment (ptoscano) [inline src/subscription_manager/cli_command/register.py:392]: regarding the text of the message: exclamation marks usually work for games, not in general messages; i'd also be more explicit on the actual reason why sub-man is erroring out, eg something like "--environments is required and it was not specified"

File: src/subscription_manager/cli_command/register.py [hunk 3]
```diff
@@ -462,9 +466,16 @@ def _get_owner_cb(self, owners):
...
+
         owner_key = None
         while not owner_key:
             owner_key = input(_("Organization: "))
             readline.clear_history()
         return owner_key
+
+    @staticmethod
+    def _is_atty():
+        return sys.stdin and sys.stdin.isatty()
```
Comment (ptoscano) [inline src/subscription_manager/cli_command/register.py:481]: this is a good idea, however the same logic is duplicated in user_pass.py; hence, i'd move it to a more generic place such as src/subscription_manager/utils.py, with a more fitting name eg "is_interactive()"

**Review by ptoscano — CHANGES_REQUESTED** [review]
it's a nice start, there are few things to improve; note i left comments in the first occurrence of each issue, so make sure to check the rest too

File: src/subscription_manager/cli_command/user_pass.py [hunk 4]
```diff
@@ -52,8 +55,23 @@ def _get_username_and_password(username, password):
...
+                    _("Error: --username and --password are required parameters and they were not provided."),
+                )
+            elif not username:
+                system_exit(
+                    os.EX_USAGE, _("Error: --username is a required parameter and it was not provided.")
+                )
+            elif not password:
+                system_exit(
+                    os.EX_USAGE, _("Error: --password is a required parameter and it was not provided.")
+                )
```
[the remark this answers is not in the document]
Comment (DuckBoss) [reply src/subscription_manager/cli_command/user_pass.py:73]: Did you mean something like this and avoiding the grouped error message?

```python
if not username:
    system_exit(
        os.EX_USAGE, _("Error: --username is a required parameter and it was not provided.")
    )
if not password:
     system_exit(
        os.EX_USAGE, _("Error: --password is a required parameter and it was not provided.")
     )
```

[the remark this answers is not in the document]
Comment (ptoscano) [reply src/subscription_manager/cli_command/user_pass.py]: yeah, exactly

File: src/subscription_manager/cli_command/register.py [hunk 5]
```diff
@@ -387,6 +388,11 @@ def _prompt_for_environment(self):
         """
         By breaking this code out, we can write cleaner tests
         """
+        if not is_interactive():
+            system_exit(
+                os.EX_USAGE, _("Error: --environments is a required parameter and it was not provided.")
```
[the remark this answers is not in the document]
Comment (DuckBoss) [reply src/subscription_manager/cli_command/register.py:393]: Thanks for your feedback. I will make those changes.

**Review by jirihnidek — APPROVED** [review]
LGTM

**Review by ptoscano — APPROVED** [review]
LGTM, thanks!

**Comment by ptoscano** [comment]
Interesting bit I got by reviewing #3150: it does fix also `OrgCommand._get_org()`, which this PR didn't fix. Can you please create a followup PR for `main` for that?

**Comment by DuckBoss** [comment]
> Interesting bit I got by reviewing #3150: it does fix also `OrgCommand._get_org()`, which this PR didn't fix. Can you please create a followup PR for `main` for that?

Thanks for finding this, I've set up a PR to address this: https://github.com/candlepin/subscription-manager/pull/3153

**Pull request merged**